### DISCUSSION TITLE
Power consumption measurement

### DISCUSSION THREAD

**Discussion by CyrusAddy** [ultralytics/ultralytics#22404] [Q&A]
Hello, I am modifying the YOLO algorithm, and I need help to how to quantify the energy savings from this modification. How do I go about this?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @CyrusAddy, thanks for starting the discussion and for your interest in Ultralytics 🚀! This is an automated response to help get things moving—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question or a methodology discussion (like energy/power comparisons), please share as much detail as possible to help the team assist you effectively:
-What YOLO version and model sizes are you using (e.g., YOLO11 family), and what exact code paths were modified?
-Your environment: OS, Python, PyTorch, CUDA/CUDNN versions, and install method
-Hardware specifics for measurement: GPU(s)/CPU(s), power limits, drivers
-Scenario: training vs. inference, dataset size/samples, batch size, precision (FP32/FP16), image size
-How you plan to measure and compare: time windows, metrics, logging, and any profiling tools or scripts used
-Representative logs or short snippets that reproduce your setup and measurements 📈

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
Great question—quantify energy by pairing a fixed benchmark with power logging, then compare Energy = mean power × wall‑clock time for baseline vs your modified model; use Ultralytics’ [Benchmark mode](https://docs.ultralytics.com/modes/benchmark/) to standardize the workload and log device power concurrently (e.g., `nvidia-smi` for NVIDIA GPUs), as outlined in our guide [How to benchmark YOLO models like YOLO11](https://www.ultralytics.com/blog/how-to-benchmark-ultralytics-yolo-models-like-yolo11) and the [Benchmarks API reference](https://docs.ultralytics.com/reference/utils/benchmarks/). Example: run `nvidia-smi --query-gpu=power.draw --format=csv,noheader,nounits --loop-ms=50 > power.csv` while you execute a standardized benchmark such as:
```bash
yolo benchmark model=yolo11n.pt data=coco128.yaml imgsz=640 device=0
```
or in Python:
```python
from ultralytics import YOLO
YOLO("yolo11n.pt").benchmark(data="coco128.yaml", imgsz=640, device=0)
```
Use the benchmark’s total elapsed time for energy (J) and divide by images processed for J/image; if you’re on Jetson, see the reproducible example in the [NVIDIA Jetson quick start](https://docs.ultralytics.com/guides/nvidia-jetson/).

**Reply by CyrusAddy** [reply]
How do you do that using a local PC? Mostly, I use the yaml file since I am training on my own dataset from scratch. How about that? Also, I am more interested in power consumption during deployment? Can I use the "predict" or "val" cli commands together with some functions/commands to capture the power consumption?

**Reply by glenn-jocher** [reply]
@CyrusAddy Yes—on a local PC you can log power while running `val` or `predict` on your dataset YAML, then compute energy as the integral of power over time; for NVIDIA GPUs, run a sampler in parallel and use your YAML as usual:

```bash
# start power logging (50 ms samples) to gpu_power.csv
nvidia-smi --query-gpu=power.draw --format=csv,noheader,nounits --loop-ms=50 > gpu_power.csv &

# validation on your dataset YAML
yolo val model=path/to/your_model.pt data=path/to/your_data.yaml imgsz=640 device=0

# or deployment-style inference on a folder/video
yolo predict model=path/to/your_model.pt source=path/to/images_or_video imgsz=640 device=0
```

If you prefer doing it in Python without external tools, you can sample system metrics (including GPU power in watts) during `predict()` and write them to CSV:

```python
import csv, time, threading
from ultralytics import YOLO
from ultralytics.utils import LOGGER

stop = False
def log_power(fp="gpu_power.csv", dt=0.05):
    with open(fp, "w", newline="") as f:
        w = csv.writer(f); w.writerow(["ts_s","gpu0_w"])
        while not stop:
            m = LOGGER.get_metrics()           # includes 'gpus': {0: {'power': W, ...}}
            w.writerow([time.time(), m["gpus"].get(0, {}).get("power")])
            f.flush(); time.sleep(dt)

t = threading.Thread(target=log_power, daemon=True); t.start()
YOLO("path/to/your_model.pt").predict(source="path/to/images_or_video", imgsz=640, device=0, stream=False)
stop = True; t.join()
```

For deployment testing, export the exact runtime you’ll ship (e.g., TensorRT or ONNX) and measure the same way:

```bash
yolo export model=path/to/your_model.pt format=engine device=0
yolo predict model=path/to/your_model.engine source=path/to/video.mp4 imgsz=640 device=0
```

The logger metrics API above is documented in the [logger.get_metrics reference](https://docs.ultralytics.com/reference/utils/logger/), and if you also want a quick compute/memory profile to explain any power deltas, see the [profile_ops reference](https://docs.ultralytics.com/reference/utils/torch_utils/).

**Reply by CyrusAddy** [reply]
@glenn-jocher Which of the following is the best for practical results: 1) inferencing on a single image, 2) inferencing on a single image for a number of iterations, and 3) inferencing on a test dataset (say 2k). And does any of these approaches change the code provided above? What its effect on the power logging?

**Reply by glenn-jocher** [reply]
For practical (deployment-like) results you’ll get the most representative and stable power numbers by running inference on a workload that matches production (typically option 3: a sizeable test set or a real video stream) long enough to smooth out warmup/IO jitter; the single-image cases are mainly useful for micro-benchmarking and should be done with a short warmup + many iterations if you want repeatable averages. The power logger/sampler code doesn’t change—only the inference workload you run changes—so the main “effect” is that longer, steadier workloads produce cleaner power traces and more reliable energy-per-image; see the runtime knobs/behavior in the [Predict mode docs](https://docs.ultralytics.com/modes/predict/) (and similarly you can use `val` for dataset-style runs).

```python
from ultralytics import YOLO

model = YOLO("path/to/your_model.pt")

# warmup (exclude from your energy/avg-power calc)
for _ in range(10):
    model.predict(source="path/to/image.jpg", imgsz=640, device=0, verbose=False)

# repeated single-image iterations (if you want option 2)
for _ in range(200):
    model.predict(source="path/to/image.jpg", imgsz=640, device=0, verbose=False)

# for option 3, just run predict on a folder/video (or val on your dataset) for a longer window
```

**Reply by RAGHAVKAKANI11** [reply]
is there any library or any way to check power consumed by the system in your script while running or knowing when benchmarking your model ?

**Reply by glenn-jocher** [reply]
Yes—inside Python you can poll NVIDIA GPU power while `predict()` or `benchmark()` runs using `GPUInfo`/NVML; the built-in stats expose `power_draw` in watts as shown in the [GPUInfo reference](https://docs.ultralytics.com/reference/utils/autodevice). Note this is GPU power, not full PC wall power, so total system power needs a wall meter or vendor telemetry.

```python
import threading, time
from ultralytics import YOLO
from ultralytics.utils.autodevice import GPUInfo

gpu = GPUInfo()
samples, stop = [], False

def poll():
    while not stop:
        gpu.refresh_stats()
        if gpu.gpu_stats:
            samples.append(gpu.gpu_stats[0]["power_draw"])  # watts
        time.sleep(0.05)

t = threading.Thread(target=poll, daemon=True)
t.start()
YOLO("yolo26n.pt").benchmark(data="coco8.yaml", imgsz=640, device=0)
stop = True
t.join()

print(f"avg GPU power: {sum(samples) / len(samples):.1f} W")
```
